packages/tpetra/core/src/Tpetra_Details_packCrsMatrix.hpp

```diff
@@ -70,17 +72,181 @@ class Distributor;
+struct PackCrsMatrixError {
+
+  size_t bad_index; // only valid if outOfBounds == true.
+  size_t bad_offset;   // only valid if outOfBounds == true.
+  size_t bad_num_bytes; // only valid if outOfBounds == true.
+  bool out_of_bounds_error;
+  bool packing_error;
+
+  PackCrsMatrixError():
+    bad_index(std::numeric_limits<size_t>::max()),
+    bad_offset(0),
+    bad_num_bytes(0),
+    out_of_bounds_error(false),
+    packing_error(false) {}
+
+  KOKKOS_INLINE_FUNCTION bool success() const
+  {
+    return ((!out_of_bounds_error) && (!packing_error));
+  }
+
+  KOKKOS_INLINE_FUNCTION std::string summary() const
+  {
+    std::ostringstream os;
+    os << "firstBadIndex: " << bad_index
+       << ", firstBadOffset: " << bad_offset
+       << ", firstBadNumBytes: " << bad_num_bytes
+       << ", outOfBounds: " << (out_of_bounds_error ? "true" : "false");
+    return os.str();
+  }
+};
+
+template<class ViewType, class OffsetsViewType,
+         class ExportsViewType, class ExportLIDsViewType,
+         class LocalMatrixType, class LocalMapType>
+struct PackCrsMatrixFunctor {
+
+  typedef typename LocalMapType::local_ordinal_type LO;
+  typedef typename LocalMapType::global_ordinal_type GO;
+  typedef typename LocalMatrixType::value_type IST;
+  typedef PackCrsMatrixError value_type;
+
+  static_assert (std::is_same<LO, typename LocalMatrixType::ordinal_type>::value,
+                 "LocalMapType::local_ordinal_type and "
+                 "LocalMatrixType::ordinal_type must be the same.");
+
+  ViewType num_packets_per_lid;
+  OffsetsViewType offsets;
+  ExportsViewType exports;
+  ExportLIDsViewType export_lids;
+  LocalMatrixType local_matrix;
+  LocalMapType local_col_map;
+
+  PackCrsMatrixFunctor(ViewType num_packets_per_lid_, OffsetsViewType offsets_,
+      ExportsViewType exports_, ExportLIDsViewType export_lids_,
+      LocalMatrixType local_matrix_, LocalMapType local_col_map_):
+    num_packets_per_lid(num_packets_per_lid_), offsets(offsets_),
+    exports(exports_), export_lids(export_lids_), local_matrix(local_matrix_),
+    local_col_map(local_col_map_)
+  {}
+
+  KOKKOS_INLINE_FUNCTION void init(value_type& dst) const
+  {
+    dst.bad_index = std::numeric_limits<size_t>::max();
+    dst.bad_offset = 0;
+    dst.bad_num_bytes = 0;
+    dst.out_of_bounds_error = false;
+    dst.packing_error = false;
+  }
+
+  KOKKOS_INLINE_FUNCTION void
+  join (volatile value_type& dst, const volatile value_type& src) const
+  {
+    if (src.bad_index < dst.bad_index) {
+      dst.bad_index = src.bad_index;
+      dst.bad_offset = src.bad_offset;
+      dst.bad_num_bytes = src.bad_num_bytes;
+      dst.out_of_bounds_error = src.out_of_bounds_error;
+      dst.packing_error = src.packing_error;
+    }
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(LO i, value_type& dst) const
+  {
+    const LO export_lid = export_lids[i];
+    const size_t buf_size = exports.size();
+
+    // FIXME (mfh 24 Mar 2017) Everything here assumes UVM.  If
```

Comment:
Remove the FIXME comment.  This is a functor now; the caller controls where it runs.